File: combinations/combinations.cpp


class Solution 
{
public:
    vector<vector<int>>ans;
    void solve(vector<int>&op,int i,int n,int k)
    {
        if(op.size()==k)
        {
            ans.push_back(op);
            return;
        }
        if(i==n+1)
        {
            return;
        }
        op.push_back(i);
        solve(op,i+1,n,k);
        op.pop_back();
        solve(op,i+1,n,k);
    }
    vector<vector<int>>combine(int n, int k)
    {
        vector<int>op;
        solve(op,1,n,k);
        return ans;
    }
};

File: 77-combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        nums = [i+1 for i in range(n)]
        ans = []
        def solve(ind, tmp):
            if len(tmp) == k :
                ans.append(list(tmp))
                return 
            if ind>=len(nums):
                return
            tmp.append(nums[ind])
            solve(ind+1, tmp)
            tmp.pop()
            solve(ind+1, tmp)
        solve(0,[])
        return ans